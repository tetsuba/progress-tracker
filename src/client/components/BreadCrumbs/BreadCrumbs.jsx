import React from 'react'
import { Breadcrumb, Row } from 'react-bootstrap'
import { Link } from 'react-router-dom'

export default function BreadCrumbs({ crumbs }) {
  function addCrumb({ path, name }) {
    return path ? (
      <li key={name} className="breadcrumb-item">
        <Link to={path}>{name}</Link>
      </li>
    ) : (
      <Breadcrumb.Item key={name} active>
        {name}
      </Breadcrumb.Item>
    )
  }

  function Crumbs() {
    return crumbs.map(addCrumb)
  }

  return (
    <Row>
      <Breadcrumb>
        <Crumbs />
      </Breadcrumb>
    </Row>
  )
}
