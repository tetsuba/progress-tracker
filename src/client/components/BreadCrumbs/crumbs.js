export const CRUMBS_KEY = {
  MY_ACCOUNT: 'myAccount',
  STUDENTS: 'Students',
  STUDENT: 'Student',
  ALPHABET: 'Alphabet',
}

const PATH = {
  HOME: { path: '/', name: 'Home' },
  MY_ACCOUNT: { path: '/myAccount', name: 'My Account' },
  STUDENTS: { path: '/students', name: 'Students' },
  STUDENT: { path: '/student/{id}', name: 'Student', id: true, replace: true },
  ALPHABET: {
    path: '/student/{id}/assessment/Alphabet',
    name: 'Alphabet',
    id: true,
  },
}

export default {
  [CRUMBS_KEY.MY_ACCOUNT]: [PATH.HOME, PATH.MY_ACCOUNT],
  [CRUMBS_KEY.STUDENTS]: [PATH.HOME, PATH.STUDENTS],
  [CRUMBS_KEY.STUDENT]: [PATH.HOME, PATH.STUDENTS, PATH.STUDENT],
  [CRUMBS_KEY.ALPHABET]: [
    PATH.HOME,
    PATH.STUDENTS,
    PATH.STUDENT,
    PATH.ALPHABET,
  ],
}
