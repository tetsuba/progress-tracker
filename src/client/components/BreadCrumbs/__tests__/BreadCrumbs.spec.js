import { testRendererFull } from '../../../../test/testHelper'
import BreadCrumbs from '../BreadCrumbs'
import CRUMBS, { CRUMBS_KEY } from '../crumbs'

jest.mock('react-router-dom', () => ({
  Link: ({ children }) => children,
}))

describe('<BreadCrumbs>', () => {
  function expectToBeALink(link, crumb) {
    expect(link.children).toEqual(crumb.name)
    expect(link.to).toEqual(crumb.path)
  }

  describe('[SCREEN]: MyAccount', () => {
    const [crumbHome, crumbMyAccount] = CRUMBS[CRUMBS_KEY.MY_ACCOUNT]

    const wrapper = testRendererFull(BreadCrumbs, {
      crumbKey: CRUMBS_KEY.MY_ACCOUNT,
    })

    it('should render crumbs "Home / My Account"', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render "Home" as a Link', () => {
      expectToBeALink(wrapper.find('Link').props(), crumbHome)
    })

    it('should render "MyAccount" as list', () => {
      expect(wrapper.find('BreadcrumbItem > li').props().children).toEqual(
        crumbMyAccount.name
      )
    })
  })
  describe('[SCREEN]: Students', () => {
    const [crumbHome, crumbStudents] = CRUMBS[CRUMBS_KEY.STUDENTS]

    const wrapper = testRendererFull(BreadCrumbs, {
      crumbKey: CRUMBS_KEY.STUDENTS,
    })

    it('should render crumbs "Home / Students"', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render "Home" as a Link', () => {
      expectToBeALink(wrapper.find('Link').props(), crumbHome)
    })

    it('should render "Students" as list', () => {
      expect(wrapper.find('BreadcrumbItem > li').props().children).toEqual(
        crumbStudents.name
      )
    })
  })
  describe('[SCREEN]: Student', () => {
    const [crumbHome, crumbStudents] = CRUMBS[CRUMBS_KEY.STUDENT]
    const props = {
      crumbKey: CRUMBS_KEY.STUDENT,
      name: 'John',
    }
    const wrapper = testRendererFull(BreadCrumbs, props)

    it('should render crumbs "Home / Students / { student name }"', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render "Home" as a Link', () => {
      expectToBeALink(wrapper.find('Link').get(0).props, crumbHome)
    })

    it('should render "Students" as a Link', () => {
      expectToBeALink(wrapper.find('Link').get(1).props, crumbStudents)
    })

    it('should render "Student" as list and replace student name with John', () => {
      expect(wrapper.find('BreadcrumbItem > li').props().children).toEqual(
        props.name
      )
    })
  })
  describe('[SCREEN]: AlphabetAssessment', () => {
    const [crumbHome, crumbStudents, crumbStudent, crumbAlphabet] = CRUMBS[
      CRUMBS_KEY.ALPHABET
    ]
    const props = {
      crumbKey: CRUMBS_KEY.ALPHABET,
      name: 'John',
      id: 'ID123456789',
    }
    const wrapper = testRendererFull(BreadCrumbs, props)

    it('should render crumbs "Home / Students / { student name }, / ABC"', () => {
      expect(wrapper).toMatchSnapshot()
    })

    it('should render "Home" as a Link', () => {
      expectToBeALink(wrapper.find('Link').get(0).props, crumbHome)
    })

    it('should render "Students" as a Link', () => {
      expectToBeALink(wrapper.find('Link').get(1).props, crumbStudents)
    })

    it('should render "Student" as Link and replace student name with John', () => {
      crumbStudent.name = props.name
      crumbStudent.path = crumbStudent.path.replace('{id}', props.id)
      expectToBeALink(wrapper.find('Link').get(2).props, crumbStudent)
    })

    it('should render "Student" as list and replace student name with John', () => {
      expect(wrapper.find('BreadcrumbItem > li').props().children).toEqual(
        crumbAlphabet.name
      )
    })
  })
})
