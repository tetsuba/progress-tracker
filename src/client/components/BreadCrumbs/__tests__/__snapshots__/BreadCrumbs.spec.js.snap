// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<BreadCrumbs> [SCREEN]: ABC should render crumbs "Home / Students / { student name }, / ABC" 1`] = `
<BreadCrumbs
  crumbKey="ABC"
  id="ID123456789"
  name="John"
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Breadcrumb
        label="breadcrumb"
        listProps={Object {}}
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            className="breadcrumb"
          >
            <Crumbs>
              <li
                className="breadcrumb-item"
                key="Home-0"
              >
                <Link
                  to="/"
                >
                  Home
                </Link>
              </li>
              <li
                className="breadcrumb-item"
                key="Students-1"
              >
                <Link
                  to="/students"
                >
                  Students
                </Link>
              </li>
              <li
                className="breadcrumb-item"
                key="John-2"
              >
                <Link
                  to="/student/ID123456789"
                >
                  John
                </Link>
              </li>
              <BreadcrumbItem
                active={true}
                key="ABC-3"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  ABC
                </li>
              </BreadcrumbItem>
            </Crumbs>
          </ol>
        </nav>
      </Breadcrumb>
    </div>
  </Row>
</BreadCrumbs>
`;

exports[`<BreadCrumbs> [SCREEN]: MyAccount should render crumbs "Home / My Account" 1`] = `
<BreadCrumbs
  crumbKey="myAccount"
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Breadcrumb
        label="breadcrumb"
        listProps={Object {}}
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            className="breadcrumb"
          >
            <Crumbs>
              <li
                className="breadcrumb-item"
                key="Home-0"
              >
                <Link
                  to="/"
                >
                  Home
                </Link>
              </li>
              <BreadcrumbItem
                active={true}
                key="My Account-1"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  My Account
                </li>
              </BreadcrumbItem>
            </Crumbs>
          </ol>
        </nav>
      </Breadcrumb>
    </div>
  </Row>
</BreadCrumbs>
`;

exports[`<BreadCrumbs> [SCREEN]: Student should render crumbs "Home / Students / { student name }" 1`] = `
<BreadCrumbs
  crumbKey="Student"
  name="John"
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Breadcrumb
        label="breadcrumb"
        listProps={Object {}}
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            className="breadcrumb"
          >
            <Crumbs>
              <li
                className="breadcrumb-item"
                key="Home-0"
              >
                <Link
                  to="/"
                >
                  Home
                </Link>
              </li>
              <li
                className="breadcrumb-item"
                key="Students-1"
              >
                <Link
                  to="/students"
                >
                  Students
                </Link>
              </li>
              <BreadcrumbItem
                active={true}
                key="John-2"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  John
                </li>
              </BreadcrumbItem>
            </Crumbs>
          </ol>
        </nav>
      </Breadcrumb>
    </div>
  </Row>
</BreadCrumbs>
`;

exports[`<BreadCrumbs> [SCREEN]: Students should render crumbs "Home / Students" 1`] = `
<BreadCrumbs
  crumbKey="Students"
>
  <Row
    noGutters={false}
  >
    <div
      className="row"
    >
      <Breadcrumb
        label="breadcrumb"
        listProps={Object {}}
      >
        <nav
          aria-label="breadcrumb"
        >
          <ol
            className="breadcrumb"
          >
            <Crumbs>
              <li
                className="breadcrumb-item"
                key="Home-0"
              >
                <Link
                  to="/"
                >
                  Home
                </Link>
              </li>
              <BreadcrumbItem
                active={true}
                key="Students-1"
                linkProps={Object {}}
              >
                <li
                  aria-current="page"
                  className="breadcrumb-item active"
                >
                  Students
                </li>
              </BreadcrumbItem>
            </Crumbs>
          </ol>
        </nav>
      </Breadcrumb>
    </div>
  </Row>
</BreadCrumbs>
`;
